src/Aspire.Dashboard/Components/Controls/ResourceDetails.razor.cs

```diff
@@ -117,111 +113,101 @@ private IEnumerable<DisplayedEndpoint> GetEndpoints()
-        else
-        {
-            // Complex values such as arrays and objects will be output as JSON.
-            // Consider how complex values are rendered in the future.
-            value = summaryValue.Value.ToString();
-        }
-        if (summaryValue.Key == KnownProperties.Container.Id)
-        {
-            // Container images have a short ID of 12 characters
-            value = value.Substring(0, Math.Min(value.Length, 12));
-        }
-        else
+        void PopulateMissingKnownProperties()
         {
-            // Dates are returned as ISO 8601 text.
-            // Use try parse to check if a value matches ISO 8601 format. If there is a match then convert to a friendly format.
-            if (DateTime.TryParseExact(value, "o", CultureInfo.InvariantCulture, DateTimeStyles.None, out var date))
+            List<KnownProperty>? knownProperties = null;
+
+            // Lazily populate some data on the view models, that wasn't available when they were constructed.
+            foreach ((var key, var vm) in Resource.Properties)
             {
-                value = FormatHelpers.FormatDateTime(timeProvider, date, cultureInfo: CultureInfo.CurrentCulture);
+                // A priority of -1 means the known property hasn't been looked up. Do that once per property, now.
+                if (vm.Priority == -1)
+                {
+                    knownProperties ??= GetKnownPropertiesForSelectedResourceType();
+
+                    var found = false;
+
+                    for (var i = 0; i < knownProperties.Count; i++)
+                    {
+                        var kp = knownProperties[i];
+
+                        if (kp.Key == key)
+                        {
+                            found = true;
+                            vm.KnownProperty = kp;
+                            vm.Priority = i;
+                            break;
+                        }
+                    }
+
+                    if (!found)
+                    {
+                        vm.Priority = int.MaxValue;
+                    }
+                }
             }
-        }
 
-        return value;
-    }
-
-    private void ToggleMaskState()
-    {
-        if (Resource.Environment is var environment)
-        {
-            foreach (var vm in environment)
+            List<KnownProperty> GetKnownPropertiesForSelectedResourceType()
             {
-                vm.IsValueMasked = _areEnvironmentVariablesMasked;
+                return Resource.ResourceType switch
+                {
+                    KnownResourceTypes.Project => _projectProperties,
+                    KnownResourceTypes.Executable => _executableProperties,
+                    KnownResourceTypes.Container => _containerProperties,
+                    _ => _resourceProperties
+                };
             }
         }
     }
 
-    private void CheckAllMaskStates()
+    private static RenderFragment GetContentAfterValue(DisplayedEndpoint vm)
```

Comment:
I'm guessing this method is where the bug with non-URL addresses is.

Creating render fragments in code behind is laborious and error prone. We should do this in the razor file where HTML can be written inline.

Reply:
Example of this: https://github.com/dotnet/aspire/blob/16953c4e71f6af49a6af671899a4eef754510580/src/Aspire.Dashboard/Components/Controls/SpanDetails.razor#L161-L194